Add Ballerina language support

xlight05 opened the pull request:
$Subject as discussed on https://github.com/microsoft/vscode-azurefunctions/issues/3361

[Ballerina](https://ballerina.io/) is a programming language designed for writing network applications. Ballerina has the support for azure functions using `ballerinax/azure_functions` package. This package gives users cleaner abstractions to write azure functions just like main languages(such as java), hiding away the complexity of using custom handlers directly. The programming model does not require users to write `functions.json` files as its automatically generated from the ballerina compiler itself. (Similar to node,python new programming model).  


I've attached some screen recordings to cover some scenarios just to make the review process easier.

1. Creating Project 

https://user-images.githubusercontent.com/23274047/222972702-bc0caa87-2862-44ca-9aac-f7da6e452656.mp4


2. Add function to existing project

https://user-images.githubusercontent.com/23274047/222972719-690c7a42-a051-4c9f-8795-ee766d1f4386.mp4


3. Initialize vscode in an existing project

https://user-images.githubusercontent.com/23274047/222972740-70a6030d-f67a-46e2-b130-bd1b43c99094.mp4


Let me know if theres any screen recordings needs to be added.
Please review the PR carefully as this is my first time contributing to this repository :) Any suggestions, requests are welcome.

xlight05 commented:
@microsoft-github-policy-service agree company="Wso2"

xlight05 commented:
@nturinski Can you take a look at this PR when you have time please?

alexweininger commented:
Where did you find the Ballerina templates? Or did you write those yourself?

xlight05 commented:
> Where did you find the Ballerina templates? Or did you write those yourself?
> 
> edit: also thanks for your contribution!

Thankss. Yes, I wrote them myself. However, the plan is to make the templates similar(at least consistent) with ballerina docs as well. 
https://ballerina.io/learn/run-in-the-cloud/function-as-a-service/azure-functions/
https://ballerina.io/learn/by-example/azure-functions-deployment/ (we'll be introducing separate examples like this for each trigger, this will align with the templates)

nturinski commented on src/commands/initProjectForVSCode/InitVSCodeStep/BallerinaInitVSCodeStep.ts:
```diff
@@ -0,0 +1,55 @@
+
+    private _debugSubpath: string = "target/azure_functions";
+
+    protected async executeCore(context: IJavaProjectWizardContext): Promise<void> {
```
Is there a `getDebugConfiguration` for this?

xlight05 replied:
I'll look into this bit futher

nturinski commented on src/utils/ballerinaUtils.ts:
```diff
@@ -0,0 +1,29 @@
+        try {
+            await cpUtils.executeCommand(undefined, undefined, ballerinaCommand, '--version');
+        } catch (error) {
+            const message: string = localize('ballerinaNotFound', 'Failed to find "bal", please ensure that the bal bin directory is in your system path.');
```
I know Java doesn't do this, but I think it'd be helpful to link the Ballerina download page in either the message or have an "open" button that comes up with the error message.

xlight05 replied:
> I know Java doesn't do this, but I think it'd be helpful to link the Ballerina download page in either the message or have an "open" button that comes up with the error message.

Great suggestion, will add this ;)

xlight05 replied:
Done

nturinski reviewed (commented):
Thanks again for your contribution and your patience! I left a few comments, but also had some broader questions:

- Would you be a good contact point if anything Ballerina ends up having issues? Most likely this will be users having issues debugging their project or fuddling with their configurations.
- Would you be willing to add tests? I can assist you with that, but it would be just the create project tests with this specific runtime.

xlight05 commented:
> Thanks again for your contribution and your patience! I left a few comments, but also had some broader questions:
> 
> * Would you be a good contact point if anything Ballerina ends up having issues? Most likely this will be users having issues debugging their project or fuddling with their configurations.
> * Would you be willing to add tests? I can assist you with that, but it would be just the create project tests with this specific runtime.

Thanks for reviewing, I'll reply and work on these suggestions.

1. For any ballerina azure functions issue, you can contact me or [@anuruddhal](https://github.com/anuruddhal). You can link our discord as well. All of the ballerina team members are present there and they will provide answers within a few hours =)

2. Yes I can definitely add tests. I found some test files related to javascript and c# but for java it was very minimal.I was not exactly sure about the test structure. If you could tell me the required files to add tests(and maybe provide a reference), I can take it from there.

I'll address other comments inline.

xlight05 commented on src/LocalResourceProvider.ts:
```diff
@@ -92,6 +92,8 @@ async function getCompiledProjectInfo(context: IActionContext, projectPath: stri
             return { compiledProjectPath: path.join(projectPath, getJavaDebugSubpath(functionAppName, buildTool)), isIsolated: false };
         }
+    } else if (projectLanguage === ProjectLanguage.Ballerina) {
+        return { compiledProjectPath: path.join(projectPath, "target", "azure_functions"), isIsolated: false };
```
It is not configurable. When the `bal build` command is executed in the ballerina project. the compiler will target the directory it contains all the outputs from the compiler. 

You can find a sample tree after executing bal build below.

.
├── Ballerina.toml
├── Dependencies.toml
├── host.json
├── httpTrigger1.bal
├── local.settings.json
├── target
│   ├── azure_functions
│   │   ├── get-httpTrigger1
│   │   │   └── function.json
│   │   ├── host.json
│   │   ├── local.settings.json
│   │   └── my_package.jar
│   ├── bin
│   │   └── my_package.jar
│   ├── build
│   └── cache
└── timerTrigger1.bal

xlight05 commented on src/commands/createNewProject/ProjectCreateStep/BallerinaProjectCreateSteps.ts:
```diff
@@ -0,0 +1,43 @@
+    protected gitignore: string = ballerinaGitIgnore;
+
+    public async executeCore(context: IBallerinaProjectWizardContext, progress: Progress<{ message?: string | undefined; increment?: number | undefined }>): Promise<void> {
+        await ballerinaUtils.validateBallerinaInstalled(context);
```
Agreed, will fix this

xlight05 commented on src/commands/createNewProject/ballerinaSteps/BallerinaPackageNameStep.ts:
```diff
@@ -0,0 +1,24 @@
+        const options: InputBoxOptions = {
+            placeHolder: localize('packageNamePlaceHolder', 'Package Name'),
+            prompt: localize('packageNamePrompt', 'Provide a name for the ballerina package'),
+            value: 'my_package'
```
thats possible. will make the changes

xlight05 replied:
Addressed this in a different way, Please check https://github.com/microsoft/vscode-azurefunctions/pull/3584#issuecomment-1537800274

xlight05 commented on src/commands/createNewProject/ballerinaSteps/BallerinaPackageVersionStep.ts:
```diff
@@ -0,0 +1,24 @@
+        const options: InputBoxOptions = {
+            placeHolder: localize('packageVersionPlaceHolder', 'Package Version'),
+            prompt: localize('packageVersionPrompt', 'Provide the version of the ballerina package'),
+            value: '1.0.0'
```
No, this is the version of the project the user creates. 

`bal --version` gives the version of the ballerina distribution installed in your pc.(similar to java 11)

xlight05 commented:
@nturinski 
I've added a [commit ](https://github.com/microsoft/vscode-azurefunctions/pull/3584/commits/32a2aa8a93f37e9259e930de1a4fbfa18b9bb413)with fixes for most of the comments and few more changes except the tests. This contains few more changes that weren't requested as well.

- Removed prompt for package name, org, and version as they are redundant. Now, upon project creation, it executes bal init in the directory which initializes a ballerina package(Ballerina.toml) in the directory with guessed package name and orgs. This removes the need to writing package name guessing logic in the vscode side. 

- Added partial debug capability. We need to figure out the workerArgKey for custom handlers. I noticed in java, you can set this  `languageWorkers__java__arguments` property as application properties and it gets picked out when the function is starting out by the func command. but for custom handlers that doesn't seem to be the case. :/. There maybe a different key for custom handlers.If you know how to do this/know anyone who does, please let me know. Made an issue as well to get this information. https://github.com/Azure/azure-functions-host/issues/9260

  However, in host.json custom handler args you can just put the debug value and then the debug server starts up.
  https://gist.github.com/xlight05/3d4e58e2d78dda827cbfb5a1c3a62348
  Is it possible to add arguments to host.json using vscode ballerina debug provider?

xlight05 commented on src/commands/createNewProject/ballerinaSteps/BallerinaPackageNameStep.ts:
```diff
@@ -0,0 +1,24 @@
+export class BalPackageNameStep extends AzureWizardPromptStep<IBallerinaProjectWizardContext> {
+    public async prompt(context: IBallerinaProjectWizardContext): Promise<void> {
+        const options: InputBoxOptions = {
+            placeHolder: localize('packageNamePlaceHolder', 'Package Name'),
```
Addressed this in a different way, Please check https://github.com/microsoft/vscode-azurefunctions/pull/3584#issuecomment-1537800274

xlight05 commented:
@nturinski Hi, Sorry, was figuring out a way to add args to startup for custom handler for providing debugger capabilities. Found a workaround for this. I'll send commit along with tests early next week. =) hope thats oky

nturinski commented on src/constants.ts:
```diff
@@ -68,6 +69,14 @@ export enum JavaBuildTool {
+    native = 'native'
+}
+
+
```
nit: extra line break

nturinski commented on src/debug/BallerinaDebugProvider.ts:
```diff
@@ -0,0 +1,32 @@
+export const defaultJavaDebugPort: number = 5005;
+
+export const ballerinaDebugConfig: DebugConfiguration = {
+    name: localize('attachJava', 'Attach to Ballerina Functions'),
```
nit: localize label needs to be changed to `attachBallerina`

nturinski commented on src/templates/script/getScriptVerifiedTemplateIds.ts:
```diff
@@ -39,7 +39,7 @@ export function getScriptVerifiedTemplateIds(version: string): (string | RegExp)
         // NOTE: The Python Preview IDs are only temporary.
         // NOTE: The Node Programming Model IDs include -4.x as a suffix
-        const regExps = verifiedTemplateIds.map(t => new RegExp(`^${t}-(JavaScript(-4.x)?|TypeScript(-4.x)?|Python|PowerShell|Custom|Python-Preview|Python-Preview-Append)$`, 'i'));
+        const regExps = verifiedTemplateIds.map(t => new RegExp(`^${t}-(JavaScript(-4.x)?|TypeScript(-4.x)?|Python|Ballerina|PowerShell|Custom|Python-Preview|Python-Preview-Append)$`, 'i'));
```
In this case, it might make more sense to create a new function similar to how `getDotnetVerifiedTemplateIds` and `getJavaVerifiedTemplateIds` does since Ballerina technically is not part of the script provider.

xlight05 replied:
I've used `TemplateType.Script` for `BallerinaProvider`. Guess its wrong then? Shall I introduce a template provider similar to `JavaTemplateProvider` for ballerina then?

nturinski replied:
I think it would be good to have your own custom template type. That way if you ever want to add more templates and have them verified, you'll only have to mess with the tests for Ballerina rather than all of the scripting template types.

xlight05 replied:
This is done. Hope its correct = )

nturinski reviewed (commented):
Thanks for your patience and effort! While waiting for the workaround, I've left smaller comments and nits. Regarding adding args, is it possible to just add these to the  `func start` command as flags and parameters? I'm not entirely sure what exactly you are trying to add.

xlight05 commented:
@nturinski 
I added some tests with https://github.com/microsoft/vscode-azurefunctions/pull/3584/commits/e33eec58f3495944e1b3413c70ea0d896cbba2b1. Tests seem to be passing in my local pc. Seems like theres some other issues as the daily build also is failing. Can you review the tests and see if they are oky?

Also kind reminder on https://github.com/microsoft/vscode-azurefunctions/pull/3584#discussion_r1204955976 

Debugger stuff seem to be oky, works fine =)

nturinski commented on test/templateCount.test.ts:
```diff
@@ -45,9 +45,13 @@ function addSuite(source: TemplateSource | undefined): void {
+            { language: ProjectLanguage.Java, version: FuncVersion.v2, expectedCount: 4 },
             // https://github.com/microsoft/vscode-azurefunctions/issues/1605
             // { language: ProjectLanguage.Java, version: FuncVersion.v3, expectedCount: 4}]
+            { language: ProjectLanguage.Ballerina, version: FuncVersion.v1, expectedCount: 5 },
```
Are you actually creating templates for all runtimes of functions? I believe v2 and v3 are deprecated or EOL, so we probably don't need to add new tests for that.

nturinski commented:
> Also, ballerina does not have a reusable task for azure devops pipelines yet, I installed ballerina manually on test workflow as a workaround, we'll make sure to implement a proper task and use it here =)

That would be great! I had a few more comments but everything is looking good :)

xlight05 commented:
> > Also, ballerina does not have a reusable task for azure devops pipelines yet, I installed ballerina manually on test workflow as a workaround, we'll make sure to implement a proper task and use it here =)
> 
> That would be great! I had a few more comments but everything is looking good :)

aye, will add changes during the weekend. 
I noticed daily build was failing for long time. will we be able to fix the issue?

nturinski commented:
We're currently trying to fix them, but unfortunately there's been so many changes with the extension, it's been hard to keep everything working.

Could you pull down from main? I tried to do it for you, but I wasn't able to pull down your branch. I think it may have been too long since there was a push?

xlight05 commented:
> We're currently trying to fix them, but unfortunately there's been so many changes with the extension, it's been hard to keep everything working.
> 
> Could you pull down from main? I tried to do it for you, but I wasn't able to pull down your branch. I think it may have been too long since there was a push?

I pushed a merge commit but daily build was failing that day too ;( however, seems like the daily build is passing today. I will push another commit to trigger the pr checks again.

I've added the ballerina template type as well to move away from the script provider.  Please go through the code one more time to see if its correct.

Besides the azure pipelines task, I think we are almost done with the implementation right? Just FYI, I've wrote a sample azure pipelines task so that we can use it for this repo workflows. 
We are planning to host the code in https://github.com/ballerina-platform/ballerina-azure-pipelines-task 
Im in the process of sending that PR, publishing it to vscode marketplace and getting it verified = )

nturinski reviewed (approved):
Looks good to me! We'll merge this soon. Thanks again for all of your patience with our collaboration! I know that I had a lot of requests. We really appreciate your contribution :)

xlight05 commented:
> Looks good to me! We'll merge this soon. Thanks again for all of your patience with our collaboration! I know that I had a lot of requests. We really appreciate your contribution :)

Great to hear!
Also, I finished the initial implementation of the DevOps pipeline task(You will still need to subscribe to the extension) - 
VS marketplace - https://marketplace.visualstudio.com/items?itemName=ballerina.ballerina-devops-pipelines
Code - https://github.com/ballerina-platform/ballerina-azure-pipelines-task/pull/1
Consumer - https://github.com/xlight05/az_pipelines_playground/blob/master/build.yml

If you can subscribe to this we can get rid of ugly ballerina installers in your workflows :D
